addons/mail/models/mail_template.py

```diff
@@ -347,7 +351,9 @@ def _generate_template_recipients(self, res_ids, render_fields,
                     additional_values['company_id'] = company_id
                 partners = self.env['res.partner']._find_or_create_from_emails(
                     all_emails,
-                    additional_values=additional_values
+                    additional_values=additional_values,
+                    default_create_values_getter=lambda norm_email:
+                    Model.browse(mail_to_res_ids[norm_email]).sudo()._get_customer_information().get(norm_email, {}),
```

Comment:
Good idea, I have used  additional_values parameter as a dict or a callable and used there as a callable which also set the company_id. I haven’t yet removed the loop to split by company because  I wonder if we do, there would be usage where customer information are leaked from one company to another. I mean if we do everything in batch, the callable only receive the email so it will gather customer info from record related to this email no matter the company. What do you think ?